pkg/vault/contracts/TimelockAuthorizer.sol

```diff
@@ -0,0 +1,368 @@
+    }
+
+    /**
+     * @dev Tells whether `account` can perform action `actionId` in `where`
+     */
+    function canPerform(
+        bytes32 actionId,
+        address account,
+        address where
+    ) public view override returns (bool) {
+        return (delaysPerActionId[actionId] > 0) ? account == address(this) : hasPermission(actionId, account, where);
+    }
+
+    /**
+     * @dev Sets a new delay for action `actionId`
+     */
+    function setDelay(bytes32 actionId, uint256 delay) external {
+        _require(msg.sender == address(this), Errors.SENDER_NOT_ALLOWED);
+        delaysPerActionId[actionId] = delay;
+        emit ActionDelaySet(actionId, delay);
+    }
+
+    /**
+     * @dev Schedules a delay change of `newDelay` for action `actionId`
+     */
+    function scheduleDelayChange(
+        bytes32 actionId,
+        uint256 newDelay,
+        address[] memory executors
+    ) external returns (uint256 scheduledActionId) {
+        require(newDelay <= MAX_DELAY, "DELAY_TOO_LARGE");
+        bytes32 scheduleDelayActionId = keccak256(abi.encodePacked(SCHEDULE_DELAY_ACTION_ID, actionId));
+        _require(hasPermission(scheduleDelayActionId, msg.sender, address(this)), Errors.SENDER_NOT_ALLOWED);
+
+        uint256 actionDelay = delaysPerActionId[actionId];
+        bytes memory data = abi.encodeWithSelector(this.setDelay.selector, actionId, newDelay);
+        return _scheduleWithDelay(scheduleDelayActionId, address(this), data, actionDelay, executors);
+    }
+
+    /**
+     * @dev Schedules a new action
+     */
+    function schedule(
+        address where,
+        bytes memory data,
+        address[] memory executors
+    ) external returns (uint256 scheduledActionId) {
+        require(where != address(this), "CANNOT_SCHEDULE_AUTHORIZER_ACTIONS");
+        bytes32 actionId = IAuthentication(where).getActionId(_decodeSelector(data));
+        _require(hasPermission(actionId, msg.sender, where), Errors.SENDER_NOT_ALLOWED);
+        return _schedule(actionId, where, data, executors);
+    }
+
+    /**
+     * @dev Executes a scheduled action `scheduledActionId`
+     */
+    function execute(uint256 scheduledActionId) external returns (bytes memory result) {
+        require(scheduledActionId < scheduledActions.length, "ACTION_DOES_NOT_EXIST");
+        ScheduledAction storage scheduledAction = scheduledActions[scheduledActionId];
+        require(!scheduledAction.executed, "ACTION_ALREADY_EXECUTED");
+        require(!scheduledAction.cancelled, "ACTION_ALREADY_CANCELLED");
+
+        // solhint-disable-next-line not-rely-on-time
+        require(block.timestamp >= scheduledAction.executableAt, "ACTION_NOT_EXECUTABLE");
+        bytes32 executeActionId = _getExecuteActionId(scheduledActionId);
+        bool isAllowed = !scheduledAction.protected || hasPermission(executeActionId, msg.sender, address(this));
+        _require(isAllowed, Errors.SENDER_NOT_ALLOWED);
```

Comment:
```suggestion
        if (scheduledAction.protected) {
            bytes32 executeActionId = _getExecuteActionId(scheduledActionId);
            _require(hasPermission(executeActionId, msg.sender, address(this)), Errors.SENDER_NOT_ALLOWED);
        }
```

Reply:
Not sure why this change

Reply:
We skip having to read the execute action id if the action is protected. But more importantly, I think the version with the `if` statement better conveys that the requirement on the caller only exists if the action is protected.